src/frontend/src/expr/utils.rs

```diff
@@ -605,6 +605,126 @@ impl WatermarkAnalyzer {
+            ExprImpl::InputRef(_) => {
+                columns.push(expr);
+            }
+            _ => (),
+        }
+    }
+
+    /// Try wrapping inner expression with `IsNotNull`
+    /// Note: only column (i.e., `InputRef`) will be extracted and connected via `AND`
+    fn try_wrap_inner_expression(expr: ExprImpl) -> Option<ExprImpl> {
+        let mut columns = vec![];
+
+        extract_column(expr, &mut columns);
+        if columns.is_empty() {
+            return None;
+        }
+
+        let mut inputs: Vec<ExprImpl> = vec![];
+        // From [`c1`, `c2`, ... , `cn`] to [`IsNotNull(c1)`, ... , `IsNotNull(cn)`]
+        for column in columns {
+            let Ok(expr) = FunctionCall::new(ExprType::IsNotNull, vec![column]) else {
+                return None;
+            };
+            inputs.push(expr.into());
+        }
+
+        // Connect them with `AND` if multiple columns are involved
+        // i.e., AND [`IsNotNull(c1)`, ... , `IsNotNull(cn)`]
+        if inputs.len() > 1 {
+            let Ok(expr) = FunctionCall::new(ExprType::And, inputs) else {
+                return None;
+            };
+            Some(expr.into())
+        } else {
+            Some(inputs[0].clone())
+        }
+    }
+
+    let ExprImpl::FunctionCall(e1_func) = e1.clone() else {
+        return (false, None);
+    };
+    let ExprImpl::FunctionCall(e2_func) = e2.clone() else {
+        return (false, None);
+    };
+    if e1_func.func_type() != ExprType::Not && e2_func.func_type() != ExprType::Not {
+        return (false, None);
+    }
+    if e1_func.func_type() != ExprType::Not {
+        if e2_func.inputs().len() != 1 {
+            return (false, None);
+        }
+        (e1 == e2_func.inputs()[0].clone(), try_wrap_inner_expression(e1))
+    } else {
+        if e1_func.inputs().len() != 1 {
+            return (false, None);
+        }
+        (e2 == e1_func.inputs()[0].clone(), try_wrap_inner_expression(e2))
+    }
+}
+
+struct SimplifyFilterExpressionRewriter {}
+
+impl ExprRewriter for SimplifyFilterExpressionRewriter {
+    fn rewrite_expr(&mut self, expr: ExprImpl) -> ExprImpl {
+        let ExprImpl::FunctionCall(func_call) = expr.clone() else {
+            return expr;
+        };
+        if func_call.func_type() != ExprType::Or && func_call.func_type() != ExprType::And {
+            return expr;
+        }
+        assert_eq!(func_call.return_type(), DataType::Boolean);
+        // Currently just optimize the first rule
+        if func_call.inputs().len() != 2 {
+            return expr;
+        }
+        let inputs = func_call.inputs();
+        let (optimizable_flag, columns) = check_pattern(inputs[0].clone(), inputs[1].clone());
+        if optimizable_flag {
+            match func_call.func_type() {
+                ExprType::Or => if let Some(columns) = columns { columns } else { ExprImpl::literal_bool(true) },
+                // `AND` will always be false, no matter the underlying columns are null or not
```

Comment:
Correct me if I'm wrong.

Reply:
It depends on the actual expression of the arms of the and.
```
select 1 = 2 and true;
 ?column?
----------
 f

select null and true;
 ?column?
----------
null

select id is null and true from t;
 ?column?
----------
 f
 f
(2 rows)

select id is not null and true from t;
 ?column?
----------
 t
 t
```